### PR TITLE
ft. ENG-34

### PR DESCRIPTION
**Pull request by Connorbelez**
## Summary by CodeRabbit

* **New Features**
  * Added share reservation functionality with idempotency support for safe retries
  * Implemented balance validation and error handling for insufficient balance and minimum fraction constraints
  * Reservation flow now locks pending fields and tracks journal entries

* **Tests**
  * Added comprehensive test suite covering happy paths, error scenarios, mutex behavior, and edge cases

* **Documentation**
  * Added implementation plan and design documentation for the reservation system

### CODE REVIEW

File: specs/ENG-34/chunks/chunk-01-reserve-shares-mutation/context.md [hunk 1]
```diff
@@ -0,0 +1,202 @@
...
+    sellerInvestorId: deal.sellerId,
+    buyerInvestorId: deal.buyerId,
+    amount: deal.fractionalShare,
+    dealId: deal._id,
+    effectiveDate: new Date(deal.closingDate).toISOString().split("T")[0],
+    idempotencyKey: `deal:${deal._id}:reserve`,
+    source: { type: "system", channel: "deal_closing" },
+  }
+);
+```
```
[the remark this answers is not in the document]
Comment (Connorbelez) [reply specs/ENG-34/chunks/chunk-01-reserve-shares-mutation/context.md]: Fixed: updated the deal-closing contract example to use sellerLenderId and buyerLenderId so it matches reserveSharesArgsValidator and the current code.

File: convex/ledger/mutations.ts [hunk 2]
```diff
@@ -258,3 +259,85 @@ export const redeemShares = ledgerMutation
...
+			metadata: args.metadata,
+		});
+
+		const amountDelta = BigInt(args.amount);
+		await ctx.db.patch(sellerPosition._id, {
+			pendingCredits: sellerPosition.pendingCredits + amountDelta,
+		});
+		await ctx.db.patch(buyerPosition._id, {
+			pendingDebits: buyerPosition.pendingDebits + amountDelta,
+		});
```
[the remark this answers is not in the document]
Comment (Connorbelez) [reply convex/ledger/mutations.ts]: Fixed: reserveShares now re-reads the seller and buyer accounts after postEntry before applying pending field patches, so the lock math no longer depends on a stale pre-postEntry snapshot.

File: convex/ledger/__tests__/reservation.test.ts [hunk 3]
```diff
@@ -0,0 +1,402 @@
...
+type ReserveSharesMutation = {
+	_handler: (
+		ctx: MutationCtx,
+		args: ReserveSharesArgs
+	) => Promise<ReserveSharesResult>;
+};
+
+const reserveSharesMutation = reserveShares as unknown as ReserveSharesMutation;
+
+function getConvexErrorCode(error: unknown): string {
```
[the remark this answers is not in the document]
Comment (Connorbelez) [reply convex/ledger/__tests__/reservation.test.ts]: Fixed: removed the hand-maintained reserveShares arg typing in tests, tightened the direct-call helper around the validator shape, and kept the handler path only because Convex codegen is blocked in this worktree by missing CONVEX_DEPLOYMENT. The test still exercises the real mutation logic in a deterministic way.

File: convex/ledger/__tests__/reservation.test.ts [hunk 4]
```diff
@@ -0,0 +1,402 @@
...
+}
+
+async function getJournalEntry(
+	t: ReturnType<typeof createTestHarness>,
+	journalEntryId: Id<"ledger_journal_entries">
+) {
+	return t.run(async (ctx) => ctx.db.get(journalEntryId));
+}
+
+describe("reserveShares", () => {
```
[the remark this answers is not in the document]
Comment (Connorbelez) [reply convex/ledger/__tests__/reservation.test.ts]: Fixed: added edge-case coverage for same-lender reservations and the missing-seller path so the contract is explicit and protected by tests.

File: convex/ledger/mutations.ts [hunk 5]
```diff
@@ -258,3 +259,85 @@ export const redeemShares = ledgerMutation
 		});
 	})
 	.public();
+
+// ── Tier 3: Two-Phase Reservation ─────────────────────────────────
+
+export const reserveShares = internalMutation({
```
[the remark this answers is not in the document]
Comment (Connorbelez) [reply convex/ledger/mutations.ts]: Verified: the checked-in generated internal API already supports this mutation. reservation.test.ts now calls internal.ledger.mutations.reserveShares directly, and both `bun typecheck` and `bunx vitest run convex/ledger/__tests__/reservation.test.ts` pass without any _generated changes in this worktree. `bunx convex codegen` is still blocked here by missing CONVEX_DEPLOYMENT, but that blocker is no longer preventing callers/tests from using reserveShares.
